src/txrequest.cpp

```diff
@@ -0,0 +1,592 @@
+        auto peerit = m_peerinfo.find(it->m_peer);
+        peerit->second.m_completed -= it->m_state == State::COMPLETED;
+        peerit->second.m_requested -= it->m_state == State::REQUESTED;
+        m_index.get<Tag>().modify(it, std::move(modifier));
+        peerit->second.m_completed += it->m_state == State::COMPLETED;
+        peerit->second.m_requested += it->m_state == State::REQUESTED;
+    }
+
+    //! Convert a CANDIDATE_DELAYED announcement into a CANDIDATE_READY. If this makes it the new best
+    //! CANDIDATE_READY (and no REQUESTED exists) and better than the CANDIDATE_BEST (if any), it becomes the new
+    //! CANDIDATE_BEST.
+    void PromoteCandidateReady(Iter<ByTxHash> it)
+    {
+        assert(it != m_index.get<ByTxHash>().end());
+        assert(it->m_state == State::CANDIDATE_DELAYED);
+        // Convert CANDIDATE_DELAYED to CANDIDATE_READY first.
+        Modify<ByTxHash>(it, [](Announcement& ann){ ann.m_state = State::CANDIDATE_READY; });
+        // The following code relies on the fact that the ByTxHash is sorted by txhash, and then by state (first
+        // _DELAYED, then _READY, then _BEST/REQUESTED). Within the _READY announcements, the best one (highest
+        // priority) comes last. Thus, if an existing _BEST exists for the same txhash that this announcement may
+        // be preferred over, it must immediately follow the newly created _READY.
+        auto it_next = std::next(it);
+        if (it_next == m_index.get<ByTxHash>().end() || it_next->m_txhash != it->m_txhash ||
+            it_next->m_state == State::COMPLETED) {
+            // This is the new best CANDIDATE_READY, and there is no IsSelected() announcement for this txhash
+            // already.
+            Modify<ByTxHash>(it, [](Announcement& ann){ ann.m_state = State::CANDIDATE_BEST; });
+        } else if (it_next->m_state == State::CANDIDATE_BEST) {
+            Priority priority_old = m_computer(*it_next);
+            Priority priority_new = m_computer(*it);
+            if (priority_new > priority_old) {
+                // There is a CANDIDATE_BEST announcement already, but this one is better.
+                Modify<ByTxHash>(it_next, [](Announcement& ann){ ann.m_state = State::CANDIDATE_READY; });
+                Modify<ByTxHash>(it, [](Announcement& ann){ ann.m_state = State::CANDIDATE_BEST; });
+            }
+        }
+    }
+
+    //! Change the state of an announcement to something non-IsSelected(). If it was IsSelected(), the next best
+    //! announcement will be marked CANDIDATE_BEST.
+    void ChangeAndReselect(Iter<ByTxHash> it, State new_state)
+    {
+        assert(new_state == State::COMPLETED || new_state == State::CANDIDATE_DELAYED);
+        assert(it != m_index.get<ByTxHash>().end());
+        if (it->IsSelected() && it != m_index.get<ByTxHash>().begin()) {
+            auto it_prev = std::prev(it);
+            // The next best CANDIDATE_READY, if any, immediately preceeds the REQUESTED or CANDIDATE_BEST
```

Comment:
d9dc98b73d3d3b7979e76514e822e74814f8ff6d, typo:
```suggestion
            // The next best CANDIDATE_READY, if any, immediately precedes the REQUESTED or CANDIDATE_BEST
```

Reply:
Done.